pageserver/src/virtual_file/owned_buffers_io/write/flush.rs

```diff
@@ -0,0 +1,200 @@
+pub struct FlushHandleInner<Buf, W> {
+    /// A bi-directional channel that sends (buffer, offset) for writes,
+    /// and receives recyled buffer.
+    channel: Duplex<(FullSlice<Buf>, u64), FullSlice<Buf>>,
+    /// Join handle for the background flush task.
+    join_handle: tokio::task::JoinHandle<std::io::Result<Arc<W>>>,
+}
+
+/// A handle to the flush task.
+pub struct FlushHandle<Buf, W> {
+    inner: Option<FlushHandleInner<Buf, W>>,
+    /// Immutable buffer for serving tail reads.
+    /// `None` if no flush request has been submitted.
+    pub(super) maybe_flushed: Option<Buf>,
+}
+
+/// A background task for flushing data to disk.
+pub struct FlushBackgroundTask<Buf, W> {
+    /// A bi-directional channel that receives (buffer, offset) for writes,
+    /// and send back recycled buffer.
+    channel: Duplex<FullSlice<Buf>, (FullSlice<Buf>, u64)>,
+    /// A writter for persisting data to disk.
+    writer: Arc<W>,
+    ctx: RequestContext,
+}
+
+impl<Buf, W> FlushBackgroundTask<Buf, W>
+where
+    Buf: IoBufAligned + Send + Sync,
+    W: OwnedAsyncWriter + Sync + 'static,
+{
+    /// Creates a new background flush task.
+    fn new(
+        channel: Duplex<FullSlice<Buf>, (FullSlice<Buf>, u64)>,
+        file: Arc<W>,
+        ctx: RequestContext,
+    ) -> Self {
+        FlushBackgroundTask {
+            channel,
+            writer: file,
+            ctx,
+        }
+    }
+
+    /// Runs the background flush task.
+    async fn run(mut self, slice: FullSlice<Buf>) -> std::io::Result<Arc<W>> {
+        // Sends the extra buffer back to the handle.
+        self.channel.send(slice).await.map_err(|_| {
+            std::io::Error::new(std::io::ErrorKind::BrokenPipe, "flush handle closed early")
+        })?;
+
+        //  Exit condition: channel is closed and there is no remaining buffer to be flushed
+        while let Some((slice, offset)) = self.channel.recv().await {
+            // Write slice to disk at `offset`.
+            let slice = self.writer.write_all_at(slice, offset, &self.ctx).await?;
+
+            // Sends the buffer back to the handle for reuse. The handle is in charged of cleaning the buffer.
+            if self.channel.send(slice).await.is_err() {
+                // Although channel is closed. Still need to finish flushing the remaining buffers.
+                continue;
+            }
+        }
+
+        Ok(self.writer)
+    }
+}
+
+impl<Buf, W> FlushHandle<Buf, W>
+where
+    Buf: IoBufAligned + Send + Sync + Clone,
+    W: OwnedAsyncWriter + Send + Sync + 'static + std::fmt::Debug,
+{
+    /// Spawns a new background flush task and obtains a handle.
+    pub fn spawn_new<B>(file: Arc<W>, buf: B, ctx: RequestContext) -> Self
+    where
+        B: Buffer<IoBuf = Buf> + Send + 'static,
+    {
+        let (front, back) = duplex_channel(2);
+
+        let bg = FlushBackgroundTask::new(back, file, ctx);
+        let join_handle = tokio::spawn(async move { bg.run(buf.flush()).await });
```

Comment:
The whole dance with passing in the `FullSlice` returned by `buf.flush()` into `bg.run()`, just so that `bg.run()` can send it back into the channel is weird (and under-commented).
Would make more sense to fill the `front` directly in `spawn_new`, I think.

also, a nitpick: reading order: we don't have delare-before-use in Rust - I personally prefer this high-level code first, `FlushBackgroundTask` is a detail.

Reply:
I added some docs in 0f63c957a683f8771c8fc5bc30d2db9b29e4a58e. The reason we send the buffer back into the channel is that so we do not need to maintain a separate buffer queue.